fish-rust/src/ast.rs

```diff
@@ -0,0 +1,5206 @@
+        unsafe fn range(self: &KeywordIn) -> SourceRange;
+        unsafe fn range(self: &SemiNl) -> SourceRange;
+        unsafe fn range(self: &JobConjunctionDecorator) -> SourceRange;
+        unsafe fn range(self: &TokenBackground) -> SourceRange;
+        unsafe fn range(self: &KeywordTime) -> SourceRange;
+        unsafe fn range(self: &TokenRedirection) -> SourceRange;
+        unsafe fn range(self: &String_) -> SourceRange;
+        unsafe fn range(self: &Argument) -> SourceRange;
+    }
+}
+
+impl Ast {
+    fn extras_ffi(self: &Ast) -> Box<ExtrasFFI<'_>> {
+        Box::new(ExtrasFFI(&self.extras))
+    }
+}
+
+struct ExtrasFFI<'a>(&'a Extras);
+
+impl<'a> ExtrasFFI<'a> {
+    fn comments(&self) -> &'a [SourceRange] {
+        &self.0.comments
+    }
+    fn semis(&self) -> &'a [SourceRange] {
+        &self.0.semis
+    }
+    fn errors(&self) -> &'a [SourceRange] {
+        &self.0.errors
+    }
+}
+
+unsafe impl ExternType for Ast {
+    type Id = type_id!("Ast");
+    type Kind = cxx::kind::Opaque;
+}
+
+impl Ast {
+    fn top_ffi(&self) -> Box<NodeFFI> {
+        Box::new(NodeFFI::new(&*self.top))
+    }
+    fn dump_ffi(&self, orig: &CxxWString) -> UniquePtr<CxxWString> {
+        self.dump(&orig.from_ffi()).to_ffi()
+    }
+}
+
+fn ast_parse_ffi(src: &CxxWString, flags: u8, errors: *mut ParseErrorList) -> Box<Ast> {
+    let mut out_errors: Option<ParseErrorList> = if errors.is_null() {
+        None
+    } else {
+        Some(unsafe { &*errors }.clone())
+    };
+    let ast = Box::new(Ast::parse(
+        &src.from_ffi(),
+        ParseTreeFlags(flags),
+        &mut out_errors,
+    ));
+    if let Some(out_errors) = out_errors {
+        unsafe { *errors = out_errors };
+    }
+    ast
+}
+
+fn ast_parse_argument_list_ffi(
+    src: &CxxWString,
+    flags: u8,
+    errors: *mut ParseErrorList,
+) -> Box<Ast> {
+    let mut out_errors: Option<ParseErrorList> = if errors.is_null() {
+        None
+    } else {
+        Some(unsafe { &*errors }.clone())
+    };
+    let ast = Box::new(Ast::parse_argument_list(
+        &src.from_ffi(),
+        ParseTreeFlags(flags),
+        &mut out_errors,
+    ));
+    if let Some(out_errors) = out_errors {
+        unsafe { *errors = out_errors };
+    }
+    ast
+}
+
+fn new_ast_traversal<'a>(root: &'a NodeFFI<'a>) -> Box<Traversal<'a>> {
+    Box::new(Traversal::new(root.as_node()))
+}
+
+impl<'a> Traversal<'a> {
+    fn next_ffi(&mut self) -> Box<NodeFFI<'a>> {
+        Box::new(match self.next() {
+            Some(node) => NodeFFI::Reference(node),
+            None => NodeFFI::None,
+        })
+    }
+}
+
+impl TokenConjunction {
+    fn token_type(&self) -> ParseTokenType {
+        (self as &dyn Token).token_type()
+    }
+}
+
+impl Statement {
+    fn contents(&self) -> &StatementVariant {
+        &self.contents
+    }
+}
+
+// pub struct NodeFFI<'a>(pub Option<&'a dyn Node>);
+
+#[derive(Clone)]
+pub enum NodeFFI<'a> {
```

Comment:
Minor nit: Have we settled on all-caps FFI? Rust generally uses `NodeFfi` as the naming convention even for abbreviations.